[#211][#207] initial demo setup with declarative approach and springboot approach + command dispatching/handling (#4755)
AxonFramework/AxonFramework: examples/university-multi-tenancy-examples/README.md

# Axoniq Framework Multi-Tenancy Demos

> A runnable demonstration of Axoniq Framework 5.3's multi-tenancy support, shown two ways: wired through

Reviewer: ```suggestion
A runnable demonstration of Axoniq Framework's multi-tenancy support, shown two ways: wired through
```

A platform hosts several universities. Each is its own tenant, and their data must never mix.
Multi-tenancy lets you register a tenant-scoped component once and have the framework inject the right
tenant's instance into each message handler, so a handler never resolves a tenant itself. The
[core module](university-multi-tenancy-core/README.md) holds that code. The demo shows, at the moment:

* **Tenant-scoped component injection** into command and query handlers, each component matched by
  type, with the tenant resolved from the message metadata.
* **The tenant lifecycle**: tenants known at startup, a tenant added at runtime, an unknown tenant
  rejected, a tenant removed (closing its instances), and cleanup on shutdown.
* **A configuration-time guardrail**: registering two providers for one component type is refused 
  because the framework cannot know which instance a parameter of that type should receive.
* **Context filtering** (Axon Server): tenants are discovered from Axon Server's contexts, with the
  `_admin` context filtered out so it never becomes a tenant.
* **The disable switch** (Spring Boot): setting `axon.multitenancy.enabled=false` turns the whole
  feature off.

> * **Two configuration styles** produce the same result: the declarative Configuration API and Spring

Reviewer 2: also covers sending commands via the tenant connector
```suggestion
* **Tenant-scoped command distribution** dispatching commands via the tenants own Axon Server context
* **Two configuration styles** produce the same result: the declarative Configuration API and Spring
```

Reviewer 2: already resovled it

  [Spring Boot demo](university-multi-tenancy-springboot/README.md). It runs against Axon Server.

Anything that runs against Axon Server needs the setup in the [Axon Server](#axon-server) section
below. If you are just here to see multi-tenancy work, the declarative demo in memory is the place to
start.

## Axon Server

The Axon Server paths above share one setup, done here once. Each tenant is its own Axon Server
context, and multiple contexts are an Enterprise feature: without a valid license the server
runs a standalone trial that cannot create the per-tenant contexts the demos provision (it rejects them
with `AXONIQ-1700 Maximum number of contexts reached`).

There are two ways to license the server. Pick the one you have and start matching Docker Compose
profile from this directory. Both expose the dashboard at <http://localhost:8024>. 

### Option 1: license file

Use this if you have an Axon Server license file that allows running disconnected from Axoniq Platform.

1. Place your license file next to this README as `axon-server.license` (it is git-ignored).
2. Run `docker compose --profile license up -d`. The compose file mounts the license into the server.

> 3. Verify if axon-server is running with the correct enterprise license by visiting <http://localhost:8024/utilities/license> in your browser.

Reviewer: We talk about "Axon Server" everywhere else, so why do "axon-server" here?
```suggestion
3. Verify if Axon Server is running with the correct enterprise license by visiting <http://localhost:8024/utilities/license> in your browser.
```

### Option 2: Axoniq Platform authentication token

Use this if you do not have a license file but do have an Axoniq Platform account. The server connects
to Axoniq Platform, which provides and renews its license automatically. Find your token on the Axon
Server download page:
<https://platform.axoniq.io/resource-center/install/server/download/?version=latest>

1. Put the token in a git-ignored `.env` file next to this README:
   ```properties
   AXONIQ_PLATFORM_AUTHENTICATION="<your-token>"
   ```
   (Or export `AXONIQ_PLATFORM_AUTHENTICATION` in your shell.)
2. Run `docker compose --profile token up -d`.

> 3. Verify if axon-server is running with the correct enterprise license by visiting <http://localhost:8024/utilities/license> in your browser.

Reviewer: ```suggestion
3. Verify if Axon Server is running with the correct enterprise license by visiting <http://localhost:8024/utilities/license> in your browser.
```

Both runnable modules drive the same `DemoLifecycle` from the core, so they prove the same behavior
and differ only in how the application is configured.

## Running

Choose the configuration style you want to learn and follow that module's README for the exact steps.

* Prefer the **declarative Configuration API**? Use the
  [declarative demo](university-multi-tenancy-declarative/README.md). It runs in memory with no
  infrastructure, which is the quickest way to see the feature, and can also run against Axon Server.
* Prefer **Spring Boot autoconfiguration**? Use the
  [Spring Boot demo](university-multi-tenancy-springboot/README.md). It runs against Axon Server.

Anything that runs against Axon Server needs the setup in the [Axon Server](#axon-server) section
below. If you are just here to see multi-tenancy work, the declarative demo in memory is the place to
start.

## Axon Server

The Axon Server paths above share one setup, done here once. Each tenant is its own Axon Server

> context, and multiple contexts are an Enterprise Edition feature: without a valid license the server

Author: ```suggestion
context, and multiple contexts are an Enterprise feature: without a valid license the server
```

runs a standalone trial that cannot create the per-tenant contexts the demos provision (it rejects them
with `AXONIQ-1700 Maximum number of contexts reached`).

There are two ways to license the server. Pick the one you have and start matching Docker Compose

> profile from this directory. Both use the Enterprise Edition image and expose the dashboard at
> <http://localhost:8024>. 
> 

Reviewer 2: there is no separation in "Standard Edition" and "Enterprise Edition" anymore since 2023.1, that can be removed throughout the README and compose file (see https://docs.axoniq.io/axon-server-reference/v2026.0/axon-server/installation/local-installation/)

```suggestion
There are two ways to license the server. Pick the one you have and start matching Docker Compose
profile from this directory. Both expose the dashboard at
<http://localhost:8024>.
```
